File: static/cogs/handling.js
import { validateLocation } from "./cogs.js";
import { removeTimes } from "./cogs.js";
import { selectSeat } from "./cogs.js";

var cities = [];
export var seatIds = [];
export var trainId = 0;

// Skickar en GET request till get_citites i handling.py
fetch('/skolarbete/programmering-1/platsbokning/get_cities', {
    method: 'GET'
})
.then(response => {
    if (!response.ok) {
        throw new Error('Network response was not ok');
    }
    return response.json()
})
.then(city_data => {
    cities = city_data;
})
.catch(error => {
    console.error('Fetch error', error);
});

  /**
   * Skickar en POST request till handling.py som returnerar alla
   * tillgängliga tider och lägger till dem i <select> elementet.
   */
export function updateAvailableTimes() {
    var departure = document.getElementById('departure').value;
    var destination = document.getElementById('destination').value;
    var date = document.getElementById('date').value;
    var selectTime = document.getElementById('time');
    var { valid, alertMessage } = validateLocation(departure, destination, cities);
    
    var data = {
        departure: departure,
        destination: destination,
        date: date
    };
    
    fetch('/skolarbete/programmering-1/platsbokning/update_times', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
    })
    .then(response => {
        if (!response.ok) {
            throw new Error('Network response was not ok');
        }
        return response.json();
    })
    .then(data => {
        if (valid) {
            var availableTimes = data.availableTimes;
            selectTime.innerHTML = '';
            if (destination != '' && departure != '') {
                for (var i = 0; i < availableTimes.length; i++) {
                    var option = document.createElement('option');
                    option.value = availableTimes[i];
                    option.text = availableTimes[i];
                    selectTime.appendChild(option);
                }
                if (availableTimes.length == 0) {
                    var noTimes = document.createElement('option');
                    noTimes.value = 'N/A';
                    noTimes.text = 'Inga tider hittades';
                    selectTime.appendChild(noTimes);
                }
            }
        } else {
            removeTimes(selectTime);
            alert(alertMessage);
        }
    })
    .catch(error => {
        console.error('Fetch error:', error)
    });
}

  /**
   * Skickar en POST request till handling.py som returnerar tillgänglighet för alla säten
   * som sen blir tillagda som knappar.
   */
export function openSeats() {
    const departure = document.getElementById('departure').value;
    const destination = document.getElementById('destination').value;
    const date = document.getElementById('date').value;
    var seats = document.getElementById('seats');
    var seatButton = document.getElementById('get_seats');
    const time = document.getElementById('time').value;
    const group1 = document.getElementById('group1');
    const group2 = document.getElementById('group2');

    var data = {
        departure: departure,
        destination: destination,
        date: date,
        time: time
    };
    
    fetch('/skolarbete/programmering-1/platsbokning/get_seats', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
    })
    .then(response => {
        if (!response.ok) {
            throw new Error("Network response was not ok");
        }
        return response.json()
    })
    .then(seatData => {
        seatData = JSON.stringify(seatData);
        seatData = JSON.parse(seatData);
        var seatDataLength = Object.keys(seatData).length - 5;
        trainId = seatData['train_id']
        
        for (let i = 0; i <= seatDataLength; i++) {
            var button = document.createElement('button');
            button.disabled = seatData[i];
            button.id = i;
            button.addEventListener('click', selectSeat);
            if (i <= 15) {
                group1.appendChild(button);
            } else {
                group2.appendChild(button);
            }
        }

        if (time) {
            seats.style.display = 'block';
            seatButton.style.display = 'none';
        }

    })
    .catch(error => {
        console.error('Fetch error: ', error)
    })
}

  /**
   * Gömmer alla säteknappar och visar 'Välj säte' knappen.
   */
export function closeSeats() {
    const seats = document.getElementById('seats');
    const seatButton = document.getElementById('get_seats');
    const group1 = document.getElementById('group1');
    const group2 = document.getElementById('group2');

    while (group1.children[0]) {
        group1.removeChild(group1.lastChild);
    }
    while (group2.children[0]) {
        group2.removeChild(group2.lastChild);
    }

    if (time) {
        seats.style.display = 'none';
        seatButton.style.display = '';
    }
}

  /**
   * Skickar en POST request till handling.py med all data för platsbokningen som skapar
   * en biljett och ändrar sätet till upptagen status.
   */
export function makeTicket() {
    const centerDiv = document.getElementById('center');
    const successDiv = document.getElementById('success');
    const departure = document.getElementById('departure').value;
    const destination = document.getElementById('destination').value;
    const date = document.getElementById('date').value;
    const time = document.getElementById('time').value; 

    var data = {
        departure: departure,
        destination: destination,
        date: date,
        time: time,
        seat_ids: seatIds,
        train_id: trainId
    }
    
    fetch('/skolarbete/programmering-1/platsbokning/make_ticket', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
    })
    .then(response => {
        if (!response.ok) {
            throw new Error("Network response was not ok");
        }
        return response.json()
    })
    .then(response_data => {
        var ticket_data = response_data['ticket_data'];
        var lines = ticket_data.split('\n');
        var tickets = [];
        
        for (let i = 0; i < lines.length; i += 6) {
            var ticket = lines.slice(i, i + 6).join('\n');
            tickets.push(ticket);
        }

        tickets.forEach(function(ticketText, index) {
            var ticketDiv = document.createElement('div');
            ticketDiv.classList.add('ticket');
            ticketDiv.innerHTML = ticketText.replace(/\n/g, "<br><br>");
            successDiv.appendChild(ticketDiv);
        });        
        centerDiv.style.display = 'none';
        successDiv.style.display = 'block';
    })
    .catch(error => {
        console.error('Fetch error:', error);
    })
}

  /**
   * Skickar en POST request till handling.py som läser och 
   * returnerar biljetten där den sedan blir utskriven.
   */
export function printTicket() {
    var ticketId = prompt('ANGE DIN BILJETT ID:');

    fetch('/skolarbete/programmering-1/platsbokning/print_ticket', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: ticketId
    })
    .then(response => {
        if (!response.ok) {
            throw new Error("Network response was not ok")
        }
        return response.json()
    })
    .then(ticketData => {
        if (ticketData['statusCode'] == '204') {
            alert('ANGE EN GILTIG BILJETT ID.');
        } else if (ticketData['statusCode'] == '200') {
            var printWindow = window.open('', '_blank');
            printWindow.document.open();
            
            var formattedTicketData = ticketData['ticketData'].replace(/\n/g, '<br>');
            
            printWindow.document.write('<html><body style="font-size: 30px;">' + formattedTicketData + '</body></html');
            printWindow.document.close();
            printWindow.print();
        }
        
    })
    .catch(error => {
        console.error('Fetch error: ', error)
    });
}

  /**
   * Skickar en POST request till handling.py som avbokar
   * biljetten och ändrar sätet(erna) till tillgängliga.
   */
export function cancelTicket() {
    var ticketId = prompt('ANGE DIN BILJETT ID:');

    fetch('/skolarbete/programmering-1/platsbokning/cancel_ticket', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: ticketId
    })
    .then(response => {
        if (!response.ok) {
            throw new Error("Network response was not ok")
        }
        return response.json()
    })
    .then(responseData => {
        if (responseData['statusCode'] == 204) {
            alert('ANGE EN GILTIG BILJETT ID.');
        } else if (responseData['statusCode'] == 200) {
            alert(responseData['status']);
        }
    })
    .catch(error => {
        console.error('Fetch error: ', error)
    });
} 

File: static/cogs/handling.py
import constants
import json
import os
import random
from urllib.parse import unquote
from flask_cors import CORS
from flask import (
    Flask,
    request
)

class LogicHandler():
    """
    Hanterar bokningen, avbokningen, och lagringen av säten
    """
    def __init__(self):
        self.seat_file = open(file=constants.SEATS_FILE, mode='r+', encoding='utf-8')
        self.seats = json.load(self.seat_file)

        self.cities_file = open(file=constants.CITIES_FILE, mode='r', encoding='utf-8')
        self.cities = json.load(self.cities_file)

    app=Flask(__name__,
                static_folder=constants.STATIC_PATH,
                template_folder=constants.TEMPLATES_PATH)
    
    CORS(app)

    def update_times(self):
        """
        Hämtar tiderna för tågresan och uppdaterar listan
        """
        departure = request.json.get('departure').title()
        destination = request.json.get('destination').title()
        date = request.json.get('date')

        available_times = []

        for train in self.seats.keys():
            train_data = self.seats.get(train, {})
            empty_departures = 0

            for time, time_data in train_data.items():
                train_departure = time_data.get('departure', '').title()
                train_destination = time_data.get('destination', '').title()
                train_date = time_data.get('date', '')

                if all([train_departure == departure, train_destination == destination, train_date == date]):
                    for time in train_data:
                        if time_data.get(time) == '':
                            empty_departures += 1
                        available_times.append(time) if time not in available_times else print(time)

            if empty_departures >= 3:
                response_data = {
                        'availableTimes': sorted(available_times),
                         'train': train
                }
                return response_data

        if not len(available_times):
            for train in self.seats.keys():
                train_data = self.seats.get(train, {})
                all_departures_empty = all(time_data.get('departure', '') == '' for time_data in train_data.values())

                if all_departures_empty:
                    response_data = {
                            'availableTimes': sorted(train_data.keys()),
                            'train': train
                    }
                    return response_data

        response_data = {
                'availableTimes': sorted(available_times),
                'train': train
        }
        return response_data

    def get_seats(self):
        """
        Hämtar datan för säterna
        """
        data = request.get_json()
        time = data.get('time')

        seat_data = None

        for train in self.seats.keys():
            train_data = self.seats.get(train, {})

            for train_time in train_data.keys():
                if train_time == time:
                    seat_data = train_data[train_time]
                    break
            if seat_data:
                break

        if seat_data:
            return seat_data

    def get_cities(self):
        """
        Hämtar datan för städerna/stationerna tillgängliga
        """
        response_data = self.cities.get('cities')
        return response_data

    def make_ticket(self):
        """
        Skapar en biljett och lagrar den i datasetet
        """
        data = request.get_json()
        departure = data.get('departure', '').title()
        destination = data.get('destination', '').title()
        date = data.get('date')
        time = data.get('time')
        seat_ids = data.get('seat_ids')
        train_id = data.get('train_id')

        for train_data in self.seats.values():
            for _, time_data in train_data.items():
                if (time_data.get('train_id') == train_id):
                    for seat_id in seat_ids:
                        if seat_id in time_data:
                            time_data['departure'] = departure
                            time_data['destination'] = destination
                            time_data['date'] = date
                            time_data[seat_id] = True
                    self.seat_file.seek(0)
                    self.seat_file.truncate(0)
                    self.seat_file.close()
                    with open(constants.SEATS_FILE, 'r+') as f:
                        json.dump(self.seats, f, indent=2)
                    for seat_id in seat_ids:
                        with open(file=f'{constants.TICKETS_PATH}/ticket_{str(seat_id)+str(train_id)}.txt', mode='w', encoding='utf-8') as ticket:
                            ticket.write(f'SJ TÅGBILJETT - ID: {random.randint(0, 61771520)}\n')
                            ticket.write(f'{departure.upper()} - {destination.upper()}\n')
                            ticket.write(f'{date} {time}\n')
                            ticket.write(f'TÅG NR. {train_id} - PLATS {seat_id}\n')
                            ticket.write(f'{"RÖKARE" if int(seat_id) <= 15 else "ICKE RÖKARE"}\n')
                            ticket.write(f'{"MITTGÅNG SÄTE" if (int(seat_id) <= 15 and int(seat_id) % 2 != 0) or (int(seat_id) > 15 and int(seat_id) % 2 == 0) else "FÖNSTER SÄTE"}')
                            print(f'{ticket.name} has successfully been made.')
                    return data
        return data

    def get_ticket_data(self):
        """
        Hämtar biljett data utifrån biljett ID
        """
        data = request.get_json()
        seat_ids = data.get('seat_ids')
        train_id = data.get('train_id')
        tickets = []
        for seat_id in seat_ids:
            with open(file=f'{constants.TICKETS_PATH}/ticket_{seat_id}{train_id}.txt', mode='r', encoding='utf-8') as ticket_file:
                ticket = ticket_file.read()
                tickets.append(ticket)
        ticket_data = '\n'.join(tickets)
        return ticket_data
    
    def print_ticket(self):
        """
        Skriver ut biljetten utifrån biljett ID
        """
        dir = f'./{constants.TICKETS_PATH}'
        id = unquote(request.get_data())
    
        ticket_data = None
        for ticket_name in os.listdir(dir):
            with open(file=f'{constants.TICKETS_PATH}/{ticket_name}', mode='r', encoding='utf-8') as ticket:
                ticket_data = ticket.read()
                ticket_id = ticket_data.split("\n")[0].split(":")[1].strip()
                if id == ticket_id:
                    break
                else:
                    ticket_data = ''
    
        if ticket_data:
            return { 'ticketData': ticket_data, 'statusCode': 200 }
        else:
            return { 'ticketData': ticket_data, 'statusCode': 204 }

        
    def cancel_ticket(self):
        """
        Avbokar biljetten utifrån biljett ID
        """
        dir = f'./{constants.TICKETS_PATH}'
        id = unquote(request.get_data())
        status = ''

        for ticket_name in os.listdir(dir):
            with open(file=f'{constants.TICKETS_PATH}/{ticket_name}', mode='r', encoding='utf-8') as ticket:
                ticket_data = ticket.read().split('\n')
                ticket_id = ticket_data[0].split(':')[1].strip()
                if id == ticket_id:
                    train_id = ticket_data[3].split(' ')[2].strip()
                    seat_id = ticket_data[3].split(' ')[5].strip()

                    for train_data in self.seats.values():
                        for train_time, time_data in train_data.items():
                            if (time_data.get('train_id') == int(train_id)):
                                if seat_id in time_data:
                                    time_data[seat_id] = False
                                self.seat_file.seek(0)
                                self.seat_file.truncate(0)
                                self.seat_file.close()
                                with open(constants.SEATS_FILE, 'r+') as f:
                                    json.dump(self.seats, f, indent=2)

                    os.remove(f'./{constants.TICKETS_PATH}/{ticket_name}')
                    status = f'AVBOKADE BILJETT MED ID {id}'
                    break

        if status != '':
            return {'status': status, 'statusCode': 200}
        else:
            return {'statusCode': 204}